src/request.ts

```diff
@@ -151,83 +153,79 @@ class Request {
-            }
-          }
+        if (accessToken !== '') {
+          options.headers.authorization = `Bearer ${accessToken}`;
+        }
+      }
 
-          const proxyUrl = process.env.HTTP_PROXY || process.env.HTTPS_PROXY;
-          if (proxyUrl) {
-            options.proxy = this.createProxyConfigFromUrl(proxyUrl);
-          }
-          return this.req(options);
-        })
-        .then((res: any): void => {
-          if (resolve) {
-            resolve((options.responseType === 'stream' || options.fullResponse) ? res : res.data);
-          }
-          else {
-            const end = process.hrtime.bigint();
-            timings.api.push(Number(end - start));
+      const proxyUrl = process.env.HTTP_PROXY || process.env.HTTPS_PROXY;
 
-            _resolve((options.responseType === 'stream' || options.fullResponse) ? res : res.data);
-          }
-        }, (error: AxiosError): void => {
-          if (error && error.response &&
-            (error.response.status === 429 ||
-              error.response.status === 503)) {
-            let retryAfter: number = parseInt(error.response.headers['retry-after'] || '10');
-            if (isNaN(retryAfter)) {
-              retryAfter = 10;
-            }
-            if (this._debug) {
-              (this._logger as Logger).log(`Request throttled. Waiting ${retryAfter}sec before retrying...`);
-            }
-            setTimeout(() => {
-              this.execute(options, resolve || _resolve, reject || _reject);
-            }, retryAfter * 1000);
-          }
-          else {
-            if (reject) {
-              reject(error);
-            }
-            else {
-              const end = process.hrtime.bigint();
-              timings.api.push(Number(end - start));
-
-              _reject(error);
-            }
-          }
-        });
-    });
+      if (proxyUrl) {
+        options.proxy = this.createProxyConfigFromUrl(proxyUrl);
+      }
+
+      const res = await this.req(options);
+
+      const end = process.hrtime.bigint();
+      timings.api.push(Number(end - start));
+
+      return options.responseType === 'stream' || options.fullResponse ?
+        res as any :
+        res.data;
+    }
+    catch (error: any) {
+      const end = process.hrtime.bigint();
+      timings.api.push(Number(end - start));
+
+      if (error && error.response && (error.response.status === 429 || error.response.status === 503)) {
+        let retryAfter: number = parseInt(error.response.headers['retry-after'] || '10');
+
+        if (isNaN(retryAfter)) {
+          retryAfter = 10;
+        }
+
+        if (this._debug) {
+          await (this._logger as Logger).log(`Request throttled. Waiting ${retryAfter} sec before retrying...`);
+        }
+
+        await this.setTimeout(retryAfter);
+        return this.execute(options);
+      }
+
+      throw error;
+    }
+  }
+
+  private async setTimeout(value: number): Promise<void> {
```

Comment:
Ah, this is smart indeed, should have thought of that as well, if we can't stub it directly, stub it through a proxy. 😀

One change though: I think it would be better if we move this to a separate file. How about we create a timersUtil file for this? It will be a short file, but it makes more sense than having it in here in the request class.